Document memcached configuration for caching. (#8373)
grafana/loki: docs/sources/operations/caching.md

---
title: Caching 
menuTitle: Caching 
description: Enable and configure memcached for caching. 
weight: 100
keywords:
  - memcached
  - caching
---
<!-- For more information about how to populate front matter, see [Topic front matter]({{< relref "../../front-matter/" >}}). -->

# Caching

Loki supports caching of index writes and lookups, chunks and query results to
speed up query performance. This sections describes the recommended Memcached
configuration to enable caching for chunks and query results. The index lookup
cache is configured to be in-memory by default.

## Before you begin

- It is recommended to deploy four dedicated Memcached clusters.

> - As of 2023-02-01 the `memcached:1.6.17-alpine` is recommended.

Reviewer: ```suggestion
- As of 2023-02-01, the `memcached:1.6.17-alpine` version of the library is recommended.
```

- Consult the Loki ksonnet [memcached](https://github.com/grafana/loki/blob/main/production/ksonnet/loki/memcached.libsonnet) deployment and the ksonnet [memcached library](https://github.com/grafana/jsonnet-libs/tree/master/memcached).

## Steps

To enable and configure Memcached:

1. Deploy each Memcached service with at least three replicas and configure
   each as follows:
    1. Chunk cache `--memory-limit=4096 --max-item-size=2m --conn-limit=1024`
    1. Query result cache `--memory-limit=1024 --max-item-size=5m --conn-limit=1024`

1. Configure Loki to use the cache.
    1. If the Helm chart is used
        1. set `memcached.chunk_cache.host` to the Memecache address of for the chunk cache.
        1. set `memcached.results_cache.host` to the Memecache address of for the query result cache.
	1. set `memcached.chunk_cache.enabled=true` and `memcached.results_cache.enabled=true`.
        1. Ensure that the connection limit of Memcached is at least `number_of_clients? * max_idle_conns`.

>   1. If the Loki configuration is used

Reviewer: I think something is off with the numbering formatting here, I think this should be nested at the same level as "If the Helm chart is used" under step 2?

Author: Yes! Thank you!

## Steps

To enable and configure Memcached:

1. Deploy each Memcached service with at least three replicas and configure
   each as follows:
    1. Chunk cache `--memory-limit=4096 --max-item-size=2m --conn-limit=1024`
    1. Query result cache `--memory-limit=1024 --max-item-size=5m --conn-limit=1024`

1. Configure Loki to use the cache.
    1. If the Helm chart is used
        1. set `memcached.chunk_cache.host` to the Memecache address of for the chunk cache.
        1. set `memcached.results_cache.host` to the Memecache address of for the query result cache.
	1. set `memcached.chunk_cache.enabled=true` and `memcached.results_cache.enabled=true`.
        1. Ensure that the connection limit of Memcached is at least `number_of_clients? * max_idle_conns`.
  1. If the Loki configuration is used
      1. Configure the chunk cache
      ```yaml
      chunk_store_config:
        chunk_cache_config:
          memcached:
            batch_size: 256
            parallelism: 10
          memcached_client:
            host: <memcached host>
            addresses: ???

>             service: ???

Author: @trevorwhitney what should addresses and service be?

Author: So both, `host` and `service` must be specified?

Reviewer 2: I believe so. We use `memcached-client` in cloud so I'd stick with that as the default.

Reviewer 3: > From the docs addresses is experimental, I would suggest using the host and service fields.
> 
> host: I have pointed this to the k8s service that is deployed for memcached service: Is the port name on that service exposing memcached
> 
> I have deployed 4 x memcached
> 
> ```
> helm upgrade --install chunk-cache -n loki bitnami/memcached -f memcached-overrides.yaml
> helm upgrade --install write-cache -n loki bitnami/memcached -f memcached-overrides.yaml
> helm upgrade --install results-cache -n loki bitnami/memcached -f memcached-overrides.yaml
> helm upgrade --install index-cache -n loki bitnami/memcached -f memcached-overrides.yaml
> ```
> 
> ```
> loki:
>   memcached:
>     chunk_cache:
>       enabled: true
>       host: chunk-cache-memcached.loki.svc
>       service: memcache
>       batch_size: 256
>       parallelism: 10
>     results_cache:
>       enabled: true
>       host: results-cache-memcached.loki.svc
>       service: memcache
>       default_validity: 12h
> 
>   structuredConfig:
>     chunk_store_config:
>       write_dedupe_cache_config:
>         memcached:
>           batch_size: 100
>           parallelism: 100
>         memcached_client:
>           host: write-cache-memcached.loki.svc
>           service: memcache
>           consistent_hash: true
> 
>     storage_config:
>       index_queries_cache_config:
>         memcached:
>           batch_size: 100
>           parallelism: 100
>         memcached_client:
>           host: index-cache-memcached.loki.svc
>           service: memcache
>           consistent_hash: true
> ```

Any chance you can  document this? `memcached-overrides.yaml` 

The bitnami chart being referenced can run standalone, or run in HA mode.  But HA mode would appear to require being able to pass in multiple hosts values to Loki.  So I assume you really want it to run in standalone mode, but then we only have 1 replica, not 3.

Any additional insight would be great.

weight: 100
keywords:
  - memcached
  - caching
---
<!-- For more information about how to populate front matter, see [Topic front matter]({{< relref "../../front-matter/" >}}). -->

# Enable and configure Memcached

Loki supports caching of index writes and lookups, chunks and query results to
speed up query performance. This sections describes the recommended Memcached
configuration to enable caching.

## Before you begin

- It is recommended to deploy four dedicated Memcached clusters.
- As of 2023-02-01 the `memcached:1.6.17-alpine` is recommended.
- Consult the Loki ksonnet [memcached](https://github.com/grafana/loki/blob/main/production/ksonnet/loki/memcached.libsonnet) deployment and the ksonnet [memcached library](https://github.com/grafana/jsonnet-libs/tree/master/memcached).

## Steps

To enable and configure Memcached:

1. Deploy each Memcached service with at least three replicas and configure
   each as follows:
  1. Chunk cache `--memory-limit=4096 --max-item-size=2m --conn-limit=1024`
  1. Index lookup cache `--memory-limit=1024 --max-item-size=5m --conn-limit=1024`
  1. Index write cache `--memory-limit=1024 --max-item-size=1m --conn-limit=1024`

>   1. Query result cache `--memory-limit=1024 --max-item-size=5m --conn-limit=1024`

Author: @salvacorts I saw that you edited the Jsonnet files last. Is there anything we are missing?

1. Configure Loki to use the cache.
    1. If the Helm chart is used

>         1. set `memcached.chunk_cache.host` to the Memecache address of for the chunk cache.

Reviewer 2: we also need to set `memcached.chunk_cache.enabled` and `memcached.results_cache.enabled` both to `true.

>     1. set `memcached.results_cache.host` to the Memecache address of for the query result cache.

Author: @trevorwhitney should we add support for the index caches as well?

Author: Seems the fifocache should suffice.

    1. Ensure that the connection limit of Memcached is at least `number_of_clients? * max_idle_conns`.

>   1. If the Loki configuration is used

Author: The following defaults are from our ksonnet deployment. Should we use these in Helm, @trevorwhitney?

Author: Seems the fifocache should suffice.

Reviewer 2: @jeschkies we already have these same defaults for the chunks and results cache. Currently the helm chart does not set up a `write_dedupe_cache`, so we would have to add that section completely including the defaults.

As for the index query cache we might need to revisit that. For `write` pods there's no need, as it [defaults to an in memory cache](https://github.com/grafana/loki/blob/9920a27f06b9a808bcd0d7323a55041bc343a8bc/pkg/loki/modules.go#L523). However the store is also use by `read` pods, and I'm not sure I understand the significance of the index query cache in query flows. Is in memory sufficient? We may need to add a section for that to the helm chart.

menuTitle: Caching 
description: Enable and configure memcached for caching. 
weight: 100
keywords:
  - memcached
  - caching
---

# Caching

Loki supports caching of index writes and lookups, chunks and query results to
speed up query performance. This sections describes the recommended Memcached
configuration to enable caching for chunks and query results. The index lookup
cache is configured to be in-memory by default.

## Before you begin

- It is recommended to deploy four dedicated Memcached clusters.
- As of 2023-02-01, the `memcached:1.6.17-alpine` version of the library is recommended.
- Consult the Loki ksonnet [memcached](https://github.com/grafana/loki/blob/main/production/ksonnet/loki/memcached.libsonnet) deployment and the ksonnet [memcached library](https://github.com/grafana/jsonnet-libs/tree/master/memcached).

## Steps

To enable and configure Memcached:

1. Deploy each Memcached service with at least three replicas and configure
   each as follows:
    1. Chunk cache 
       ```
       --memory-limit=4096 --max-item-size=2m --conn-limit=1024
       ```
    1. Query result cache
       ```
       --memory-limit=1024 --max-item-size=5m --conn-limit=1024

>        ```

Reviewer 4: If we refer to 4 recommended memcached clusters here, we should specify the configs for those others too

1. Configure Loki to use the cache.
    1. If the Helm chart is used

>        Set `memcached.chunk_cache.host` to the Memecache address of for the chunk cache, `memcached.results_cache.host` to the Memecache address of for the query result cache, `memcached.chunk_cache.enabled=true` and `memcached.results_cache.enabled=true`. 

Reviewer: ```suggestion
       Set `memcached.chunk_cache.host` to the Memecache address for the chunk cache, `memcached.results_cache.host` to the Memecache address for the query result cache, `memcached.chunk_cache.enabled=true` and `memcached.results_cache.enabled=true`. 
```

       
       Ensure that the connection limit of Memcached is at least `number_of_clients * max_idle_conns`.
       

>        The options `host` and `service` depend on the type of installation. For example using the `bitname/memcached` Helm Charts with the following commands the `service` values is always `memcache`.

Reviewer: ```suggestion
       The options `host` and `service` depend on the type of installation. For example, using the `bitname/memcached` Helm Charts with the following commands, the `service` values are always `memcached`.
```
